libkbfs/interfaces.go

```diff
@@ -1319,29 +1335,56 @@ type MDServer interface {
 	// user has permission to do so.
 	//
+	// If lockBeforeGet is not nil, it takes a lock on the lock ID before
+	// trying to get anything. If taking the lock fails, an error is returned.
+	// Note that taking a lock from the mdserver is idempotent.
+	//
 	// If there is no returned error, then the returned ID must
 	// always be non-null. A nil *RootMetadataSigned may be
 	// returned, but if it is non-nil, then its ID must match the
 	// returned ID.
 	GetForHandle(ctx context.Context, handle tlf.Handle,
-		mStatus MergeStatus) (tlf.ID, *RootMetadataSigned, error)
+		mStatus MergeStatus, lockBeforeGet *keybase1.LockID) (
+		tlf.ID, *RootMetadataSigned, error)
 
 	// GetForTLF returns the current (signed/encrypted) metadata object
 	// corresponding to the given top-level folder, if the logged-in
 	// user has read permission on the folder.
-	GetForTLF(ctx context.Context, id tlf.ID, bid BranchID, mStatus MergeStatus) (
-		*RootMetadataSigned, error)
+	//
+	// If lockBeforeGet is not nil, it takes a lock on the lock ID before
+	// trying to get anything. If taking the lock fails, an error is returned.
+	// Note that taking a lock from the mdserver is idempotent.
+	GetForTLF(ctx context.Context, id tlf.ID, bid BranchID, mStatus MergeStatus,
+		lockBeforeGet *keybase1.LockID) (*RootMetadataSigned, error)
 
 	// GetRange returns a range of (signed/encrypted) metadata objects
 	// corresponding to the passed revision numbers (inclusive).
+	//
+	// If lockBeforeGet is not nil, it takes a lock on the lock ID before
+	// trying to get anything. If taking the lock fails, an error is returned.
+	// Note that taking a lock from the mdserver is idempotent.
 	GetRange(ctx context.Context, id tlf.ID, bid BranchID, mStatus MergeStatus,
-		start, stop kbfsmd.Revision) ([]*RootMetadataSigned, error)
+		start, stop kbfsmd.Revision, lockBeforeGet *keybase1.LockID) (
+		[]*RootMetadataSigned, error)
 
 	// Put stores the (signed/encrypted) metadata object for the given
 	// top-level folder. Note: If the unmerged bit is set in the metadata
 	// block's flags bitmask it will be appended to the unmerged per-device
 	// history.
-	Put(ctx context.Context, rmds *RootMetadataSigned, extra ExtraMetadata) error
+	//
+	// If lockContext is not nil, it causes the mdserver to check a lockID at
+	// the time of the put, and optionally (if specified in lockContext)
+	// releases the lock on the lock ID if the put is successful. Releasing the
+	// lock in mdserver is idempotent.
+	Put(ctx context.Context, rmds *RootMetadataSigned, extra ExtraMetadata,
+		lockContext *keybase1.LockContext, priority keybase1.MDPriority) error
+
+	// LockOps performs a locking operation with a mdserver. If isTake is true, it
+	// ensures lockID for tlfID is taken by this device, i.e., idempotently
+	// take the lock. If isTake is false, it ensures lockID for tlfID is not taken
+	// by this device, i.e., idmpotently release the lock.
+	LockOp(ctx context.Context,
```

Comment:
I have a slight preference for explicit `Lock` and `Unlock` operations, so we're not relying on a bool.  But I don't feel that strongly about it.

Reply:
OK; will do. I was just being lazy, especially thinking that we might not even use it given the locking parameters in Get/Put.